core/src/service/sync_service.rs

```diff
@@ -24,122 +28,251 @@ pub struct WorkCalculated {
+            base: (&self.tx.base_metadata).to_staged(Vec::new()),
+            local: (&self.tx.local_metadata).to_staged(Vec::new()),
+            username_by_public_key: &mut self.tx.username_by_public_key,
+            public_key_by_username: &mut self.tx.public_key_by_username,
+            client: self.client,
+            config: self.config,
+        };
+        let update_as_of = sync_context.sync(&mut |op| work_units.extend(get_work_units(&op)))?;
+
+        Ok(WorkCalculated { work_units, most_recent_update_from_server: update_as_of as u64 })
+    }
+
     #[instrument(level = "debug", skip_all, err(Debug))]
     pub fn sync<F: Fn(SyncProgress)>(
         &mut self, maybe_update_sync_progress: Option<F>,
-    ) -> CoreResult<()> {
-        // initialize sync progress: 3 metadata pulls + 1 metadata push + num local doc changes
-        // note: num doc changes can change as a result of pull (can make new/changes docs deleted or add new docs from merge conflicts)
-        let mut num_doc_changes = 0;
-        for (id, local_change) in self.tx.local_metadata.get_all() {
-            if let Some(base_file) = self.tx.base_metadata.get(id) {
-                if local_change.document_hmac() != base_file.document_hmac() {
-                    num_doc_changes += 1;
+    ) -> CoreResult<WorkCalculated> {
+        let mut work_units: Vec<WorkUnit> = Vec::new();
+        let mut sync_progress =
+            SyncProgress { total: 0, progress: 0, current_work_unit: ClientWorkUnit::PullMetadata };
+        if maybe_update_sync_progress.is_some() {
+            let mut sync_context = SyncContext {
+                dry_run: false,
+                account: self.get_account()?.clone(),
+                root: self.tx.root.get(&OneKey {}).cloned(),
+                last_synced: self.tx.last_synced.get(&OneKey {}).map(|s| *s as u64),
+                base: (&mut self.tx.base_metadata).to_staged(Vec::new()),
+                local: (&mut self.tx.local_metadata).to_staged(Vec::new()),
+                username_by_public_key: &mut self.tx.username_by_public_key,
+                public_key_by_username: &mut self.tx.public_key_by_username,
+                client: self.client,
+                config: self.config,
+            };
+            sync_context.sync(&mut |op| match op {
+                SyncOperation::PullMetadataStart
+                | SyncOperation::PushMetadataStart(_)
+                | SyncOperation::PullDocumentStart(_)
+                | SyncOperation::PushDocumentStart(_) => {}
+                SyncOperation::PullMetadataEnd(_)
+                | SyncOperation::PushMetadataEnd
+                | SyncOperation::PullDocumentEnd
+                | SyncOperation::PushDocumentEnd => {
+                    sync_progress.total += 1;
                 }
-            } else if local_change.document_hmac().is_some() {
-                num_doc_changes += 1;
-            }
+            })?;
```

Comment:
we calculate total sync progress using a dry run and handling operations by incrementing the total; this is _not_ the same as the length of work units in work calculated as it e.g. considers all pushed metadata as one unit